dbms/src/Core/Types.h

```diff
@@ -57,8 +57,123 @@ template <> struct TypeName<Float32> { static const char * get() { return "Float
+/// 13 reserved for TypeNumber<Int128>
 
 /// Not a data type in database, defined just for convenience.
 using Strings = std::vector<String>;
 
 }
+
+#if 1 /// __int128
+namespace DB
+{
+
+using Int128 = __int128;
+template <> constexpr bool IsNumber<Int128> = true;
+template <> struct TypeName<Int128>  { static const char * get() { return "Int128";  } };
+template <> struct TypeNumber<Int128>   { static constexpr const size_t value = 13; };
+
+}
+
+namespace std
+{
+
+template <> struct is_signed<__int128>
+{
+    static constexpr bool value = true;
+};
+
+template <> struct is_unsigned<__int128>
+{
+    static constexpr bool value = false;
+};
+
+template <> struct is_integral<__int128>
+{
+    static constexpr bool value = true;
+};
+
+template <> struct is_arithmetic<__int128>
+{
+    static constexpr bool value = true;
+};
+
+}
+#endif
+
+
+namespace DB
+{
+    /// Own FieldType for Decimal
+    template <typename T>
+    class Dec
+    {
+    public:
+        using NativeType = T;
+
+        Dec() = default;
+        Dec(Dec<T> &&) = default;
+        Dec(const Dec<T> &) = default;
+
+        Dec(const T & value_)
+        :   value(value_)
+        {}
+
+        template <typename U>
+        Dec(const Dec<U> & x)
+        :   value(x)
+        {}
+
+        constexpr Dec<T> & operator = (Dec<T> &&) = default;
+        constexpr Dec<T> & operator = (const Dec<T> &) = default;
+
+        operator T () const { return value; }
+
+        const Dec<T> & operator += (const T & x) { value += x; return *this; }
+        const Dec<T> & operator -= (const T & x) { value -= x; return *this; }
+        const Dec<T> & operator *= (const T & x) { value *= x; return *this; }
+        const Dec<T> & operator /= (const T & x) { value /= x; return *this; }
+        const Dec<T> & operator %= (const T & x) { value %= x; return *this; }
+
+    private:
+        T value;
+    };
+
+    using Dec32 = Dec<Int32>;
+    using Dec64 = Dec<Int64>;
+    using Dec128 = Dec<Int128>;
+
+    template <> struct TypeName<Dec32>   { static const char * get() { return "Dec32";   } };
+    template <> struct TypeName<Dec64>   { static const char * get() { return "Dec64";   } };
+    template <> struct TypeName<Dec128>  { static const char * get() { return "Dec128";  } };
+
+    template <> struct TypeNumber<Dec32>    { static constexpr const size_t value = 16; };
+    template <> struct TypeNumber<Dec64>    { static constexpr const size_t value = 17; };
+    template <> struct TypeNumber<Dec128>   { static constexpr const size_t value = 18; };
+
+    template <typename T>
+    inline constexpr bool decTrait() { return false; }
+    template <> constexpr bool decTrait<Dec32>() { return true; }
+    template <> constexpr bool decTrait<Dec64>() { return true; }
+    template <> constexpr bool decTrait<Dec128>() { return true; }
+
+    template <typename T>
+    inline constexpr bool decBaseTrait() { return false; }
```

Comment:
Missing comment.
(that it is true for the types that can be used as a storage type for Decimal)

Reply:
isValidStorageForDecimal